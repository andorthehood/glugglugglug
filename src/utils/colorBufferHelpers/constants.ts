export const VERTICES_PER_QUAD = 6;
export const COLOR_COMPONENTS_PER_QUAD = 3;
