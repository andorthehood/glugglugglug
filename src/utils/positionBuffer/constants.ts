export const VERTICES_PER_QUAD = 6;
