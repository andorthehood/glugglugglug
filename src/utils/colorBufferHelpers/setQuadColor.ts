import { COLOR_COMPONENTS_PER_QUAD } from "./constants";

/**
 * Writes a quad's color into a palette array, normalizing 0-255 inputs to 0-1.
 * @param palette Destination Float32Array storing quad colors.
 * @param quadIndex Index of the quad whose color should be updated.
 * @param r Red component (0-255).
 * @param g Green component (0-255).
 * @param b Blue component (0-255).
 */
export function setQuadColor(
	palette: Float32Array,
	quadIndex: number,
	r: number,
	g: number,
	b: number,
): void {
	const base = quadIndex * COLOR_COMPONENTS_PER_QUAD;
	palette[base + 0] = r / 255;
	palette[base + 1] = g / 255;
	palette[base + 2] = b / 255;
}
