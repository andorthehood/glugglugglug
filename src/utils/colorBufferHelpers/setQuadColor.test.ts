import { describe, it, expect } from "vitest";

import { COLOR_COMPONENTS_PER_QUAD } from "./constants";
import { setQuadColor } from "./setQuadColor";

function toArray(values: Float32Array) {
	return Array.from(values);
}

describe("setQuadColor", () => {
	it("normalizes 0-255 components to 0-1 and writes them to the palette", () => {
		const palette = new Float32Array(COLOR_COMPONENTS_PER_QUAD * 2);

		setQuadColor(palette, 0, 255, 0, 128);

		expect(toArray(palette)).toMatchSnapshot();
	});

	it("updates colors at the specified quad index", () => {
		const palette = new Float32Array(COLOR_COMPONENTS_PER_QUAD * 3);

		setQuadColor(palette, 1, 10, 20, 30);

		expect(toArray(palette)).toMatchSnapshot();
	});
});
