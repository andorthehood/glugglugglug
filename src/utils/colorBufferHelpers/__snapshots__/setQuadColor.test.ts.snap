// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setQuadColor > normalizes 0-255 components to 0-1 and writes them to the palette 1`] = `
[
  1,
  0,
  0.501960813999176,
  0,
  0,
  0,
]
`;

exports[`setQuadColor > updates colors at the specified quad index 1`] = `
[
  0,
  0,
  0,
  0.03921568766236305,
  0.0784313753247261,
  0.11764705926179886,
  0,
  0,
  0,
]
`;
